src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs

```diff
@@ -14,44 +14,18 @@ namespace System.Net.Http
     internal class WinInetProxyHelper
     {
         private bool _useProxy = false;
+        private bool _scriptDetectionFailed = false;
 
         public WinInetProxyHelper()
         {
-            var proxyConfig = new Interop.WinHttp.WINHTTP_CURRENT_USER_IE_PROXY_CONFIG();
+            // If auto-detect the URL for the proxy auto-config file failed, cache the information.
+            DetectScriptLocation();
 
-            try
+            // Unable to find an auto-config file.
+            // Check whether a proxy server is listed in Internet Explorer's settings.
+            if (_scriptDetectionFailed && string.IsNullOrEmpty(AutoConfigUrl))
             {
-                if (Interop.WinHttp.WinHttpGetIEProxyConfigForCurrentUser(out proxyConfig))
-                {
-                    AutoConfigUrl = Marshal.PtrToStringUni(proxyConfig.AutoConfigUrl);
-                    AutoDetect = proxyConfig.AutoDetect;
-                    Proxy = Marshal.PtrToStringUni(proxyConfig.Proxy);
-                    ProxyBypass = Marshal.PtrToStringUni(proxyConfig.ProxyBypass);
-
-                    WinHttpTraceHelper.Trace(
-                        "WinInetProxyHelper.ctor: AutoConfigUrl={0}, AutoDetect={1}, Proxy={2}, ProxyBypass={3}",
-                        AutoConfigUrl,
-                        AutoDetect,
-                        Proxy,
-                        ProxyBypass);
-                    _useProxy = true;
-                }
-                else
-                {
-                    // We match behavior of WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY and ignore errors.
-                    int lastError = Marshal.GetLastWin32Error();
-                    WinHttpTraceHelper.Trace("WinInetProxyHelper.ctor: error={0}", lastError);
-                }
-
-                WinHttpTraceHelper.Trace("WinInetProxyHelper.ctor: _useProxy={0}", _useProxy);
-            }
-
-            finally
-            {
-                // FreeHGlobal already checks for null pointer before freeing the memory.
-                Marshal.FreeHGlobal(proxyConfig.AutoConfigUrl);
-                Marshal.FreeHGlobal(proxyConfig.Proxy);
-                Marshal.FreeHGlobal(proxyConfig.ProxyBypass);
+                GetIEProxySetting();
```

Comment:
So the WinHttp-detected script location takes precedence over the user's local configuration? Is that correct? Or does WinHttp factor that in, too?

Reply:
>So the WinHttp-detected script location takes precedence over the user's local configuration? 

AutoDetect has precedence over the other settings.  If you look at the below picture, you see that there can be multiple settings for determining where the PAC file (if any) and then if there is a specific proxy server present.  Each checkbox represents what to use to find/use PAC file or specfic proxy in order of priority.  If a step fails, the fallback is usually to use the next setting (checkbox).

![image](https://user-images.githubusercontent.com/7684835/38468948-197d6f5a-3b02-11e8-8f26-49d73f6da26f.png)

So, "Automatically detect settings" is what calling WinHttpDetectAutoProxyConfigUrl() helps figure out (we should only be calling that if the checkbox is true and we only know that by calling 
WinHttpGetIEProxyConfigForCurrentUser() in the first place.

I will say, though, that AutoDetect could fail (i.e WPAD protocol fails) but that doesn't mean there is no PAC file, nor that we should stop calling WinHttpGetProxyForUrl() for each request.

Reply:
> So the WinHttp-detected script location takes precedence over the user's local configuration?

This is not the case for Windows 8.1 and newer. We will only using WinHttp-detected script location if failed to get the user's local configuration. Here is the proof:

1. We do a [`WinHttp.WinHttpOpen` with parameter `WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY`](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L716) (if no proxy specified & we need to detect it ourselves) before we create the `WinInetProxyHelper` object.
2. When we specify `WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY` flag for `WinHttp.WinHttpOpen`, it will use system and per-user proxy settings (including the Internet Explorer proxy configuration) to determine which proxy/proxies to use. (https://msdn.microsoft.com/en-us/library/windows/desktop/aa384098(v=vs.85).aspx)
3. If the above procedure failed, then we will try to create the `WinInetProxyHelper`, and do the winhttp detecting script location stuff.

The `WinHttpGetIEProxyConfigForCurrentUser` functionality duplicated with `WinHttpOpen`. However, notice that the flag `WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY` is only supported in Windows 8.1 and newer. Therefore, we still need to calling into `WinHttpGetIEProxyConfigForCurrentUser` for previous Windows version.

I do agree that on OS before Windows 8.1, we may have issues with the ordering (as David has pointed out: we only know if `Automatically detect settings` is on by calling WinHttpGetIEProxyConfigForCurrentUser() in the first place). I will change the code to correct the behavior.